File: 2-rbac-and-guards/backend/src/modules/admin/admin.controller.ts
/**
 * Controller REST cho feature Admin + Editor + Public health.
 * (EN: REST controller exposing admin, editor and public-health endpoints for the RBAC demo.)
 */
import {
    Controller,
    Get,
    UseGuards,
} from "@nestjs/common"
import {
    Public,
    Role,
    Roles,
    RolesGuard,
} from "../../common"
import {
    JwtAuthGuard,
} from "../auth"

/**
 * Ba namespace:
 *  - `/admin/dashboard` chỉ ADMIN (single-role).
 *  - `/editor/articles` ADMIN hoặc EDITOR (multi-role OR).
 *  - `/public/health` công khai qua `@Public()` (bypass JwtAuthGuard).
 * (EN: Three namespaces — admin single-role, editor multi-role OR, public bypass.)
 */
@Controller()
@UseGuards(JwtAuthGuard,
    RolesGuard)
export class AdminController {
    /**
     * ADMIN only — chứng minh single-role gate.
     * (EN: Admin-only dashboard demonstrating single-role gating.)
     *
     * @returns Thông tin dashboard giả lập (EN: mock dashboard payload).
     */
    @Roles(Role.ADMIN)
    @Get("admin/dashboard")
    getAdminDashboard(): { message: string; stats: { users: number; orders: number } } {
        return {
            message: "Chào mừng Admin vào khu vực mật!",
            stats: {
                users: 100,
                orders: 15,
            },
        }
    }

    /**
     * ADMIN hoặc EDITOR — chứng minh multi-role OR gate.
     * (EN: Allows either ADMIN or EDITOR — multi-role OR demo.)
     *
     * @returns Danh sách bài viết giả lập (EN: mock articles payload).
     */
    @Roles(Role.ADMIN,
        Role.EDITOR)
    @Get("editor/articles")
    getEditorArticles(): { message: string; articles: Array<{ id: number; title: string }> } {
        return {
            message: "Editor area — admin và editor đều xem được.",
            articles: [
                {
                    id: 1,
                    title: "Hello RBAC",
                },
                {
                    id: 2,
                    title: "Multi-role OR check",
                },
            ],
        }
    }

    /**
     * Public health endpoint — `@Public()` khiến JwtAuthGuard bỏ qua xác thực.
     * (EN: Public health endpoint — @Public() bypasses the JWT guard entirely.)
     *
     * @returns Trạng thái liveness đơn giản (EN: simple liveness payload).
     */
    @Public()
    @Get("public/health")
    getPublicHealth(): { status: string } {
        return {
            status: "ok",
        }
    }
}

File: 2-rbac-and-guards/1-java/src/main/java/com/example/rbacdemo/controller/AdminController.java
package com.example.rbacdemo.controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
public class AdminController {

    @GetMapping("/admin/dashboard")
    public Map<String, Object> getAdminDashboard() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("users", 100);
        stats.put("orders", 15);

        Map<String, Object> response = new HashMap<>();
        response.put("message", "Chào mừng Admin vào khu vực mật!");
        response.put("stats", stats);

        return response;
    }

    @GetMapping("/editor/articles")
    public Map<String, Object> getEditorArticles() {
        List<Map<String, Object>> articles = new ArrayList<>();
        
        Map<String, Object> a1 = new HashMap<>();
        a1.put("id", 1);
        a1.put("title", "Hello RBAC");
        articles.add(a1);

        Map<String, Object> a2 = new HashMap<>();
        a2.put("id", 2);
        a2.put("title", "Multi-role OR check");
        articles.add(a2);

        Map<String, Object> response = new HashMap<>();
        response.put("message", "Editor area — admin và editor đều xem được.");
        response.put("articles", articles);

        return response;
    }

    @GetMapping("/public/health")
    public Map<String, Object> getPublicHealth() {
        Map<String, Object> response = new HashMap<>();
        response.put("status", "ok");
        return response;
    }
}
